Allow function type for localePath

lucaslcode opened the pull request:
Fixes #1500

> `localePath` as a function is of the form `(locale: string, namespace: string, env: 'client' | 'server', missing: boolean) => string` returning the entire path including filename and extension. When `missing` is true, return the path for the `addPath` option of `i18next-fs-backend`, when false, return the path for the `loadPath` option. [More info at the `i18next-fs-backend` repo.](https://github.com/i18next/i18next-fs-backend/tree/master#backend-options)

See comments

Tests added ~~Wasn't sure how to add tests for the additional functionality - any guidance appreciated (maybe I should add tests for the error cases and that the function is correctly in the final config by calling it?)~~

lucaslcode commented on src/config/createConfig.ts:
```diff
@@ -124,22 +136,29 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
-    if (localePath.match(/^\.?\/public\//)) {
-      clientLocalePath = localePath.replace(/^\.?\/public/, '')
-    }
+    const clientLocalePath = typeof localePath === 'string' && localePath.match(/^\.?\/public\//)
```
Has to be const to satisfy typescript, since line 157 uses a closure.

lucaslcode replied:
Additionally, it seems to me this stripping of `/public` on the client isn't necessary, since nothing is even read on the client - I'm not totally sure though.

isaachinman replied:
Yeah, we can probably remove the stripping of `/public`. It's a remnant of when this package used to use `i18next-http-backend` by default, and used `getInitialProps`, necessitating client side loading of namespaces.

isaachinman commented:
Thanks @lucaslcode! Just wanted to give a heads up that I am travelling now, and probably won't be able to review this until sometime in January. Rest assured I will work with you to get this merged as soon as possible.

lucaslcode commented:
Thanks for the update, no worries!

On Tue, 28 Dec 2021, 06:07 Isaac Hinman, ***@***.***> wrote:

> Thanks @lucaslcode <https://github.com/lucaslcode>! Just wanted to give a
> heads up that I am travelling now, and probably won't be able to review
> this until sometime in January. Rest assured I will work with you to get
> this merged as soon as possible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/isaachinman/next-i18next/pull/1593#issuecomment-1001777844>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHUPIP3VCHONHVHQZCSYQQTUTDPTNANCNFSM5KTAU4XA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

isaachinman commented on src/config/createConfig.ts:
```diff
@@ -54,7 +54,7 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
       // https://github.com/isaachinman/next-i18next/issues/358
       //
-      if (typeof defaultNS === 'string' && typeof lng !== 'undefined') {
+      if (typeof serverLocalePath === 'string' && typeof defaultNS === 'string' && typeof lng !== 'undefined') {
```
If `typeof serverLocalePath === 'function'`, doesn't that make validation of `defaultNS` even easier? I think we should still validate in both cases.

isaachinman commented on src/config/createConfig.ts:
```diff
@@ -69,15 +69,27 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
+          addPath: path.resolve(process.cwd(), `${serverLocalePath}/${localeStructure}.missing.${localeExtension}`),
+          loadPath: path.resolve(process.cwd(), `${serverLocalePath}/${localeStructure}.${localeExtension}`),
+        }
+      } else {
```
Just to be safe:

```suggestion
      } else if (typeof serverLocalePath === 'function') {
```

isaachinman commented on src/config/createConfig.ts:
```diff
@@ -69,15 +69,27 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
       // Set server side preload (namespaces)
       //
       if (!combinedConfig.ns && typeof lng !== 'undefined') {
+        if (typeof serverLocalePath === 'function')
+          throw new Error('Must provide all namespaces in ns option if using a function as localePath')
```
```suggestion
        if (typeof serverLocalePath === 'function') {
          throw new Error('Must provide all namespaces in ns option if using a function as localePath')
        }
```

isaachinman commented on src/serverSideTranslations.ts:
```diff
@@ -85,6 +85,9 @@ export const serverSideTranslations = async (
   })
 
   if (!Array.isArray(namespacesRequired)) {
+    if (typeof localePath === 'function')
+      throw new Error('Must provide namespacesRequired to serverSideTranslations when using a function as localePath')
```
```suggestion
    if (typeof localePath === 'function') {
      throw new Error('Must provide namespacesRequired to serverSideTranslations when using a function as localePath')
    }
```

isaachinman reviewed (commented):
Thanks for your patience – finally had a chance to get to this.

Just a few comments to discuss a bit, but overall I am extremely impressed with both the quality of the code, and your ability to understand the wider repo.

Massive thanks for this contribution!

lucaslcode commented:
@isaachinman thanks for the review (and the compliment 😄). I've addressed the feedback, including removing the stripping of `/public` and therefore no longer including `env: 'server' | 'client'` in the callback parameters, since there isn't a use-case for it as far as I can tell.

isaachinman commented on README.md:
```diff
@@ -74,8 +74,8 @@ This tells `next-i18next` what your `defaultLocale` and other locales are, so th
   i18n: {
-    defaultLocale: 'en',
-    locales: ['en', 'de'],
+    defaultLocale: "en",
```
Can you omit your autoformatting on this file?

isaachinman commented:
Looks great @lucaslcode, thanks for all your help!

Weirdly this PR is not running tests in Circle – I think it's related to a CircleCI OAuth issue a couple weeks ago. Can you try rebasing against latest `master`? If that doesn't work, we may need to recreate the PR entirely.

lucaslcode commented:
I'll try a new PR
